libavfilter/sf_text2graphicsub.c

```diff
@@ -0,0 +1,492 @@
+    if (!context->size.height || !context->size.width) {
+        av_log(NULL, AV_LOG_ERROR, "A positive height and width are required to render subtitles\n");
+        return AVERROR_EXIT;
+    }
+    ass_set_frame_size(context->renderer, context->size.width, context->size.height);
+    ass_set_storage_size(context->renderer, context->size.width, context->size.height);
+
+    return 0;
+}
+
+static int config_output(AVFilterLink *outlink)
+{
+    AVFilterContext *ctx = outlink->src;
+    ASS_Context *context = ctx->priv;
+
+    outlink->w = context->size.width;
+    outlink->h = context->size.height;
+
+    return 0;
+}
+
+static int filter_frame(AVFilterLink *inlink, AVFrame *frame)
+{
+    const AVFilterContext *ctx = inlink->dst;
+    ASS_Context *context = ctx->priv;
+    AVFilterLink *outlink = ctx->outputs[0];
+    if (!context->track) {      // Init track
+        if (frame->subtitle_header)
+            context->track = ass_read_memory(context->library,
+                (char *) frame->subtitle_header->data,
+                frame->subtitle_header->size, NULL);
+        else context->track = ass_new_track(context->library);
+        if (!frame->subtitle_header) context->ass_split_context = avpriv_ass_split("");
+        else context->ass_split_context = avpriv_ass_split((char *) frame->subtitle_header->data);
+    }
+
+    for (int r = 0; r < frame->num_subtitle_areas; r++)
+    {
+        ASS_Image *image;
+        ASSDialog *dialog;
+        ASS_Track *track = context->track;
+        AVSubtitleArea *area = frame->subtitle_areas[r];
+        if (area->type == SUBTITLE_BITMAP) continue;
+
+        dialog = avpriv_ass_split_dialog(context->ass_split_context, area->ass);
+        if (track->n_events > 0)
+            ass_free_event(track, 0);
+        track->n_events = 0;
+        ass_alloc_event(track);
+        track->n_events = track->max_events = 1;
+        track->events[0].Start = frame->subtitle_timing.start_pts / (AV_TIME_BASE / 1000);
+        track->events[0].Duration = frame->subtitle_timing.duration / (AV_TIME_BASE / 1000);
+        track->events[0].Effect = av_strdup(dialog->effect);
+        track->events[0].Layer = dialog->layer;
+        track->events[0].MarginL = dialog->margin_l;
+        track->events[0].MarginR = dialog->margin_r;
+        track->events[0].MarginV = dialog->margin_v;
+        track->events[0].Name = av_strdup(dialog->name);
+        track->events[0].Text = av_strdup(dialog->text);
+        track->events[0].ReadOrder = dialog->readorder;
+        track->events[0].Style = 0;
+        for (int style = 0; style < track->n_styles; style++)
+            if (!strcmp(track->styles[style].Name, dialog->style))
+                track->events[0].Style = style;
+        track->events[0].render_priv = NULL;
+        avpriv_ass_free_dialog(&dialog);
+
+        image = ass_render_frame(context->renderer, track,
+            track->events[0].Start + track->events[0].Duration / 2, NULL);
```

Comment:
Picking half-way between start and end can have weird effects on the outputs when animations are active. 
Have you tested files with ass animations?

Reply:
Yes, as you know, I gave up on disabling animations. I've no idea how to do it.

Reply:
Will need to research a bit for the best way of doing this.

Reply:
By the way, I tried not using the midpoint (instead using the start time), it's no good. Some frames get lost. That is because some frames are made to fade in and out, so they are only visible during a interval somewhere around the midpoint. I think the midpoint is pretty good.

Reply:
I know, and I understand why you did it. Hold on I'll respond by e-mail with examples..